pkg/adsc/adsc.go

```diff
@@ -1096,51 +1102,76 @@ func (a *ADSC) GetEndpoints() map[string]*endpoint.ClusterLoadAssignment {
 	return a.eds
 }
 
-func (a *ADSC) handleMCP(gvk []string, rsc *any.Any, valBytes []byte) error {
+func (a *ADSC) handleMCP(gvk []string, resources []*any.Any) {
 	if len(gvk) != 3 {
-		return nil // Not MCP
+		return // Not MCP
 	}
 	// Generic - fill up the store
 	if a.Store == nil {
-		return nil
-	}
-	m := &mcp.Resource{}
-	err := types.UnmarshalAny(&types.Any{
-		TypeUrl: rsc.TypeUrl,
-		Value:   rsc.Value,
-	}, m)
-	if err != nil {
-		return err
+		return
 	}
-	val, err := mcpToPilot(m)
+
+	groupVersionKind := config.GroupVersionKind{Group: gvk[0], Version: gvk[1], Kind: gvk[2]}
+	existingConfigs, err := a.Store.List(groupVersionKind, "")
 	if err != nil {
-		adscLog.Warn("Invalid data ", err, " ", string(valBytes))
-		return err
+		adscLog.Warnf("Error handling received MCP config %v", err)
+		return
 	}
-	val.GroupVersionKind = config.GroupVersionKind{Group: gvk[0], Version: gvk[1], Kind: gvk[2]}
-	cfg := a.Store.Get(val.GroupVersionKind, val.Name, val.Namespace)
-	if cfg == nil {
-		_, err = a.Store.Create(*val)
+
+	var received = make(map[string]*config.Config)
+	for _, rsc := range resources {
+		m := &mcp.Resource{}
+		err := types.UnmarshalAny(&types.Any{
+			TypeUrl: rsc.TypeUrl,
+			Value:   rsc.Value,
+		}, m)
 		if err != nil {
-			return err
+			adscLog.Warnf("Error handling received MCP config %v", err)
+			continue
 		}
-	} else {
-		_, err = a.Store.Update(*val)
+		val, err := mcpToPilot(m)
 		if err != nil {
-			return err
+			adscLog.Warn("Invalid data ", err, " ", string(rsc.Value))
+			continue
 		}
-	}
-	if a.LocalCacheDir != "" {
-		strResponse, err := json.MarshalIndent(val, "  ", "  ")
-		if err != nil {
-			return err
+		received[val.Namespace+"/"+val.Name] = val
+
+		val.GroupVersionKind = config.GroupVersionKind{Group: gvk[0], Version: gvk[1], Kind: gvk[2]}
+		cfg := a.Store.Get(val.GroupVersionKind, val.Name, val.Namespace)
+		if cfg == nil {
+			_, err = a.Store.Create(*val)
+			if err != nil {
+				adscLog.Warnf("Error handling received MCP config %v", err)
+				continue
+			}
+		} else {
+			_, err = a.Store.Update(*val)
+			if err != nil {
+				adscLog.Warnf("Error handling received MCP config %v", err)
+				continue
+			}
 		}
-		err = ioutil.WriteFile(a.LocalCacheDir+"_res."+
-			val.GroupVersionKind.Kind+"."+val.Namespace+"."+val.Name+".json", strResponse, 0644)
-		if err != nil {
-			return err
+		if a.LocalCacheDir != "" {
+			strResponse, err := json.MarshalIndent(val, "  ", "  ")
+			if err != nil {
+				adscLog.Warnf("Error handling received MCP config %v", err)
+				continue
+			}
+			err = ioutil.WriteFile(a.LocalCacheDir+"_res."+
+				val.GroupVersionKind.Kind+"."+val.Namespace+"."+val.Name+".json", strResponse, 0644)
+			if err != nil {
+				adscLog.Warnf("Error handling received MCP config %v", err)
+			}
 		}
 	}
 
-	return nil
+	// remove deleted resources from cache
+	for _, config := range existingConfigs {
```

Comment:
This is not safe, if an existing config is updated/unchanged, this will delete it

Reply:
Name and namespace can't be updated, so updated configs won't be deleted.
xDS is using full push, so unchanged configs won't be deleted.

Reply:
The existingConfigs may contain unchanged/updated configs, so should move this logic before calling Create/Update

Reply:
I have added a test case to cover this scenario:  https://github.com/istio/istio/pull/29296/files#diff-1875cf3befe58910cd1c13e4a7e224e0d5e136f9cf6ff1bdea1d428a38b7eb8fR350

Reply:
I missed that you have excluded the resources received